vault/activity_log.go

```diff
@@ -2009,6 +2046,195 @@ func processClientRecord(e *activity.EntityRecord, byNamespace summaryByNamespac
+		Namespaces: make([]*activity.NamespaceRecord, 0, len(opts.byNamespace)),
+		Months:     make([]*activity.MonthRecord, 0, len(opts.byMonth)),
+	}
+	pq.Months = a.transformMonthBreakdowns(opts.byMonth)
+
+	for nsID, entry := range opts.byNamespace {
+		mountRecord := make([]*activity.MountRecord, 0, len(entry.Mounts))
+		for mountAccessor, mountData := range entry.Mounts {
+			mountRecord = append(mountRecord, &activity.MountRecord{
+				MountPath: a.mountAccessorToMountPath(mountAccessor),
+				Counts: &activity.CountsRecord{
+					EntityClients:    len(mountData.Counts.Entities),
+					NonEntityClients: int(mountData.Counts.Tokens) + len(mountData.Counts.NonEntities),
+				},
+			})
+		}
+
+		pq.Namespaces = append(pq.Namespaces, &activity.NamespaceRecord{
+			NamespaceID:     nsID,
+			Entities:        uint64(len(entry.Counts.Entities)),
+			NonEntityTokens: entry.Counts.Tokens + uint64(len(entry.Counts.NonEntities)),
+			Mounts:          mountRecord,
+		})
+	}
+	err := a.queryStore.Put(ctx, pq)
+	if err != nil {
+		a.logger.Warn("failed to store precomputed query", "error", err)
+	}
+	return nil
+}
+
+// pqOptions holds fields that will be used when creating precomputed queries
+// These fields will remain the same for every segment that a precomputed query worker is handling
+type pqOptions struct {
+	byNamespace map[string]*processByNamespace
+	byMonth     map[int64]*processMonth
+	// endTime sets the end time of the precomputed query.
+	// When invoked on schedule by the precomputedQueryWorker, this is the end of the month that just finished.
+	endTime time.Time
+	// activePeriodStart is the earliest date in our retention window
+	activePeriodStart time.Time
+	// activePeriodEnd is the latest date in our retention window.
+	// When invoked on schedule by the precomputedQueryWorker, this will be the timestamp of the most recent segment
+	// that's present in storage
+	activePeriodEnd time.Time
+}
+
+// segmentToPrecomputedQuery processes a single segment
+func (a *ActivityLog) segmentToPrecomputedQuery(ctx context.Context, segmentTime time.Time, reader SegmentReader, opts pqOptions) error {
+	hyperloglog, err := a.CreateOrFetchHyperlogLog(ctx, segmentTime)
+	if err != nil {
+		// We were unable to create or fetch the hll, but we should still
+		// continue with our precomputation
+		a.logger.Warn("unable to create or fetch hyperloglog", "start time", segmentTime, "error", err)
+	}
+
+	// Iterate through entities, adding them to the hyperloglog and the summary maps in opts
+	for {
+		entity, err := reader.ReadEntity(ctx)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			a.logger.Warn("failed to read segment", "error", err)
+			return err
+		}
+		err = a.handleEntitySegment(entity, segmentTime, hyperloglog, opts)
+		if err != nil {
+			a.logger.Warn("failed to handle entity segment", "error", err)
+			return err
+		}
+	}
+
+	// Store the hyperloglog
+	err = a.StoreHyperlogLog(ctx, segmentTime, hyperloglog)
```

Comment:
I tend to agree with @raskchanky for this, and I especially think it's valuable not to pollute the `vault` package namespace.

Reply:
Additionally, we can't move hyperloglog saving/writing into `handleEntitySegment` because it's possible that handleEntitySegment is called multiple times for the same segment. We would end up with multiple hll's for the same segment, which doesn't match the current behavior.

Reply:
Works for me!